docs/man/pg_replication_activity.1
.TH PG_REPLICATION_ACTIVITY 1 "Juli 2018" "Manual page for pg_replication_activity"
.\"
.\" 29th Juli 2018
.\" Man page author:
.\"    Sebastiaan Mannem <smannem@bol.com> <sebas@mannem.nl>
.SH NAME
.I pg_replication_activity
\- display replication activity on a replicated postgres cluster
.SH SYNOPSIS
.B pg_replication_activity
\-h[elp]
.br
.B pg_replication_activity
[options]
.SH DESCRIPTION
.I pg_replication_activity
connects to all cluster instances of the replicated cluster and display a top-like view of a 
replicated postgres cluster.
.br
It shows information like role, processed wal (MB/s) lag (MB and sec), current LSN, and config
information. pg_replication_activity is an ideal tool to analyze and visualize the operation and/or
problems of a replicated cluster.
.br
Information like:
.br
* How well is a async cluster keeping up (look at wal processing on master and standby)?
.br
* cause of replication lag?
.br
* is there a split brain?
.br
* what instance is most ahead in a split brain situation?
.br
* which instance is master?
.br
* What path is the replication stream following from master to this standby?
.SS OPTIONS
.TP
.IR \-h " | " \--help
display a help of all options
.TP
.IR \-c " | " \--connectstring
specify any connectstring to multiple hosts.
.TP
.IR \-r " | " \--role
Switch to this role after connecting.
.br
Usefull if user has no direct permissions, but been granted a role with permissions.
.TP
.IR \-C " | " \--nocolor
output in Black and White
.TP
.IR \-x " | " \--debug
More verbose on error output (only use is during development)
.SH CAVEATS
.PP
.I Question marks (?) are displayed
.br
If the output shows question marks (?) it's because you don't have enough permissions to read
all data required to display this info. Maybe adding permissions, or switchong to a role with 
enough permissions (--role) allows for displaying this info.
.PP
.I Monitoring multiple replicated clusters at once
.br
You can connect to instances of multiple replicated clusters at once. Although this can 
technically work, a lot of displayed info is deducted from the total. For example lag is calculated
with 'latest LSN' and 'lastest date' as reference. With multiple replicated clusters this would be
only one master of one replicated cluster. Standby's of other replicated clusters would typically
show a false (and huge lag). pg_replication_activity was never designed to work like this.
.PP
.I Passwords
.br
pg_replication_activity requires access to all instances it displays. Missing instances is dealt
with properly, but for example when local connection requires no password, but remote does, you
might still need to specify a password. And if you use a pgpass file, you need to specify passwords
for all instances of the replicated cluster.
.PP
.I Impact on monitorred cluster
.br
While small, this tool still has a footprint on the replicated cluster instances.
Futhermore, it has a footprint on the host running pg_replication_activity.
Therefore, it has (small, but still) an impact on (replication) performance.
Use entirely at your own risk. Analyse replication issues. Trace it, tune it and fix it. Become a wizard.
.SH EXAMPLES
.PP
.br
Display connection info of a replicated cluster
.br
* running on host1:5432 and host2:5433
.br
* connect as user postgres
.BR file .
.br
\fI% pg_replication_activity \'host=host1,host2 port=5432,5433 user=postgres\'\fR
.PP
.br
Display connection info of a replicated cluster defined as service mycoolpgservice
.br
Note: you should then specify host=host1,host2 and port=port1,port2 in the service config
.BR file .
.br
\fI% pg_replication_activity service=mycoolpgservice\fR
.PP
.SH "RETURN VALUES"
The following error values are returned:
.TP
0
no errors encountered.
.TP
1
Some error has occurred
.SH "SEE ALSO"
psql(1), top(1), pg_top(1), pg_activity(1), pg_view
.br
.SH NOTES
Please note that you specify connections to all instances of the replicated cluster, like you
specify client_failover in libpq. Specifically by specifying host= and port= Parameter Key Words in
connectstring or service, or by specifying PGHOST and PGPORT environment variablen.
You can specify the same conenctstring for psql (version >=10). The main difference is that
pg_replication_activity uses the config to deduct al hosts and conennt to all of them, where psql
would find the master and only keep that connection open.
.br
pg_replication_activity opens only one connection per cluster instance and keeps it open.
Furthermore pg_replication_activity runs queries with a low footprint, so running
pg_replication_activity is relatively inexpensive.
.SH VERSION
This manual page documents pg_replication_activity version 0.0.1.
.SH AUTHOR
.br
(c) 2018 by Sebastiaan Mannem
.br
<smannem@bol.com\- Bol.com>
.LP
Distribute freely and credit me and Bol.com,
.PP


pgreplicationactivity/pg_replication_activity.py
#!/usr/bin/env python
"""
This module is the starting point for pg_replication_activity.

module: pg_replication_activity
author: Sebastiaan Mannem <sebas@mannem.nl>
license: PostgreSQL License

Copyright (c) 2012 - 2018, S. Mannem

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose, without fee, and without a written
agreement is hereby granted, provided that the above copyright notice
and this paragraph and the following two paragraphs appear in all copies.

IN NO EVENT SHALL SEBASTIAAN MANNEM BE LIABLE TO ANY PARTY FOR
DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST
PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION,
EVEN IF SEBASTIAAN MANNEM HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.

SEBASTIAAN MANNEM SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING,
BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE. THE SOFTWARE PROVIDED HEREUNDER IS ON AN "AS IS"
BASIS, AND SEBASTIAAN MANNEM HAS NO OBLIGATIONS TO PROVIDE
MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
"""


import os
import sys
import signal
import argparse
import logging
import psycopg2
from psycopg2 import errorcodes

from pgreplicationactivity import UI, pgconnection

if os.name != 'posix':
    sys.exit("FATAL: Platform not supported.")

# Create the UI
PGAUI = UI.UI()


def get_arguments():
    """Run argparse and return arguments."""
    try:
        # Use argparse to handle devices as arguments
        description = 'htop like application for PostgreSQL replication ' + \
                      'activity monitoring.'
        parser = argparse.ArgumentParser(description=description)

        # -c / --connectstring
        parser.add_argument(
            '-c',
            '--connectstring',
            dest='connstr',
            default='',
            help='Connectstring (default: "").',
            metavar='CONNECTSTRING')
        # -r / --role
        parser.add_argument(
            '-r',
            '--role',
            dest='role',
            default=None,
            help='Role (default: "").',
            metavar='ROLE')
        # -C / --no-color
        parser.add_argument(
            '-C',
            '--no-color',
            dest='nocolor',
            action='store_true',
            help="Disable color usage.",)
        # --debug
        parser.add_argument(
            '-x',
            '--debug',
            dest='debug',
            action='store_true',
            help="Enable debug mode for traceback tracking.")
        args = parser.parse_args()
    except (argparse.ArgumentError, argparse.ArgumentTypeError) as err:
        print('pg_activity: error: %s' % str(err))
        print('Try "pg_activity --help" for more information.')
        sys.exit(1)
    return args


def main():
    """Run the main entrypoint."""
    signal.signal(signal.SIGTERM, PGAUI.signal_handler)
    args = get_arguments()
    if args.debug:
        logging.basicConfig(level=logging.DEBUG)
    else:
        logging.basicConfig(level=logging.INFO)

    try:
        connstr = try_connecting(args)
        # top part
        interval = 0
        if PGAUI.get_mode() == 'lag':
            lag_info = PGAUI.data.get_standby_info()
        # draw the flag
        flag = UI.get_flag_from_options()
        # main loop
        disp_procs = None
        # indentation
        indent = PGAUI.get_indent(flag)
        # Init curses
        PGAUI.init_curses()
        # color ?
        if args.nocolor is True:
            PGAUI.set_nocolor()
        else:
            PGAUI.set_color()
        while 1:
            PGAUI.check_window_size()
            old_pgtop_mode = PGAUI.get_mode()
            # poll process
            (disp_procs, new_lag_info) = PGAUI.poll(interval, flag, indent,
                                                    lag_info, disp_procs)
            if PGAUI.get_mode() != old_pgtop_mode:
                indent = PGAUI.get_indent(flag)
            lag_info = new_lag_info
            # get active connections
            PGAUI.set_buffer({
                'procs': disp_procs,
                'conn_string': connstr,
                'pg_version': PGAUI.data.get_pg_version(),
                'flag': flag,
                'indent': indent,
                'tps': 9,
                'active_connections': 10,
                'size_ev': 100,
                'total_size': 1000
            })
            # refresh
            PGAUI.refresh_window()
            interval = 1

    except KeyboardInterrupt as err:
        PGAUI.at_exit_curses()
        sys.exit(1)
    except Exception as err:
        PGAUI.at_exit_curses()
        # DEBUG
        if args.debug:
            logging.exception("FATAL: %s", str(err))
        sys.exit("FATAL: %s" % (str(err)))


def check_for_password_error(err):
    """Check if an error is a password error."""
    msg = str(err).strip()
    if msg.startswith("FATAL:  password authentication failed "
                      "for user"):
        return True
    if err.pgcode == errorcodes.INVALID_PASSWORD:
        return True
    if msg == "fe_sendauth: no password supplied":
        return True
    return False


def try_connecting(args):
    """Try connecting and retry with password if needed."""
    password = os.environ.get('PGPASSWORD')
    nb_try = 0
    while nb_try < 2:
        try:
            dsn = pgconnection.connstr_to_dsn(args.connstr)
            if password and nb_try > 0:
                dsn['password'] = password
            PGAUI.data = pgconnection.PGMultiConnection(dsn, args.role)
            PGAUI.data.connect()
            break
        except psycopg2.OperationalError as err:
            is_password_error = check_for_password_error(err)
            if is_password_error and nb_try < 1:
                nb_try += 1
                password = UI.ask_password()
            else:
                if args.debug:
                    logging.exception(str(err).strip())
                sys.exit("pg_activity: FATAL: %s" %
                         (UI.clean_str(str(err),)))

    return pgconnection.dsn_to_connstr({k: v for k, v in dsn.items() if k != 'password'})
